docs/src/app/(private)/experiments/custom-components.tsx

```diff
@@ -0,0 +1,87 @@
+'use client';
+import * as React from 'react';
+import { useRenderer, RenderProp } from '@base-ui-components/react/use-renderer';
+import styles from './custom-components.module.css';
+
+type Weight = 'light' | 'regular' | 'bold';
+type Size = 'small' | 'medium' | 'large';
+
+type TextState = {
+  weight: Weight;
+  size: Size;
+  excludedProp: boolean;
+};
+
+type TextProps = {
+  className: string | ((state: TextState) => string);
+  weight?: Weight;
+  render?: RenderProp<TextState>;
+  children: React.ReactNode;
+  style?: React.CSSProperties;
+  size?: Size;
+  excludedProp?: boolean;
+};
+
+const Text = React.forwardRef(
+  (props: TextProps, forwardedRef: React.ForwardedRef<HTMLElement>) => {
+    const {
+      className,
+      render,
+      style = {},
+      weight = 'regular',
+      size = 'medium',
+      // Example state prop that we exclude from the style hooks
+      excludedProp = true,
+      ...otherProps
+    } = props;
+
+    const fontWeight = {
+      light: 300,
+      regular: 400,
+      bold: 700,
+    }[weight];
+
+    const state = React.useMemo(
+      () => ({ weight, size, excludedProp }),
+      [weight, size, excludedProp],
+    );
+
+    const { renderElement } = useRenderer({
+      render: render ?? 'p',
```

Comment:
I've been wondering if we should get rid of the string fallbacks. We have to maintain a map in `defaultRenderFunctions` which increases bundle size and is not necessary given `render` only supports `HTMLAttributes<any>` now. i.e., they should just do `render ?? <p />`

Reply:
This is a great idea 💯 I also wondered why we have that defaultRenferFunction mapping. I think we can update the demo and just not support it for the external use-cases. Is there any benefit on having this as a string?

Reply:
`"p"` is slightly better than `<p />` as it's mapped to `(props) => <p {...props} />`. Since we use the longhand versions internally, it's better to have these functions defined in one place to avoid recreating them in each component. So a publicly exported string => render fn map doesn't really add to bundle size.

Reply:
It probably appeals to users that are used to `<Box as="p">`

Reply:
The problem is the map is non-exhaustive and doesn't cover every tag. I'm finding I also have to constantly add new ones. By default, `<tag />` does and works out of the box. The string doesn't seem worth it

Reply:
OK, it's a good point for external users as we won't map all the tags. I'd leave them for `useComponentRenderer` (or replace them with a direct reference to `defaultRenderFunctions`)